### ISSUE TITLE
Question: any chance to have a on_started handler in the Task class?

### ISSUE THREAD

**Issue by rhymes** [celery/celery#4110]
It would be very handy to have an on_started call back in the Task class. In addition to after_return, on_failure, on_success and on_retry.

Thoughts?

**Comment by georgepsarakis** [comment]
@rhymes could you describe a use case?

**Comment by rhymes** [comment]
Celery provides a [task_track_started](http://docs.celeryproject.org/en/latest/userguide/configuration.html#std:setting-task_track_started) configuration option, this allows up to open up flower and see which of the many long running tasks have been actually "STARTED" instead of just "PENDING".

We would like to show in the frontend that a task has been actually picked up and started in the queue but there's no callback for that.

As the documentation says:

> Having a ‘started’ state can be useful for when there are long running tasks and there’s a need to report what task is currently running.

But how can we report, or send notifications or tell the users in the frontend which of the many batch calculations in the queue is currently running, if there's no callback?

Hope this makes sense.

**Comment by aydin** [comment]
I don't know why `on_started` handler doesn't exist while others does.
Below structure can be workaround.
`class BaseTask(Task):
    def __call__(self, *args, **kwargs):
        self.before_run()
        return super(BaseTask, self).__call__(*args, **kwargs)
    def before_run(self):
        #do stuff
        pass

@app.task(base=BaseTask)
def task1():
    pass`

**Comment by pavankumarkatakam** [comment]
I too have a use case for `on_started` if possible `on_progress` (custom state).
The workaround specified by @aydin will so but the task state will not update right?

**Comment by rhymes** [comment]
@pavankumarkatakam no, you need to update manually the state. See here an example http://docs.celeryproject.org/en/latest/userguide/tasks.html#custom-states

**Comment by Krismix1** [comment]
To other people who find this issue & code suggestion from @aydin, so that you don't spend much time debugging.

As of recent-ish updates to Celery docs following #5652, it is recommended to call `self.run` instead of `super.__call__` when overriding the `__call__` method. We have hit a similar scenario during testing when using `__call__` as described in #5643 where tasks are not enqueued for retries.

**Comment by auvipy** [comment]
> To other people who find this issue & code suggestion from @aydin, so that you don't spend much time debugging.
> 
> As of recent-ish updates to Celery docs following #5652, it is recommended to call `self.run` instead of `super.__call__` when overriding the `__call__` method. We have hit a similar scenario during testing when using `__call__` as described in #5643 where tasks are not enqueued for retries.

so should we close this issue?

**Comment by thedrow** [comment]
I think we can add this to the documentation or add the hook if everyone wants it.

**Comment by Krismix1** [comment]
Seeing that overriding `__call__` needs to be done in a specific way, it would feel more natural and safer to have a separate handler function for the event.
I tried to add the hook myself, but I am not sure if I am on the right track, as the hook is not called in the test. [Here](https://github.com/Krismix1/celery/commit/3da2e53a8667125959d6408690020f09522cab7c) is what I have tried. Am I missing something obvious?

Otherwise, if there are not enough people in need of an explicit handler function, I guess the time could be spent better on other improvements.

**Comment by thedrow** [comment]
@Krismix1 I'd accept this as a PR if you remove the breakpoint and add documentation.

**Comment by Krismix1** [comment]
@thedrow created a PR (#6923). Please let me know if I should make any other changes.
